flocker/node/agents/blockdevice.py

```diff
@@ -0,0 +1,313 @@
+    """
+    A block device that may be attached to a host.
+
+    :ivar unicode blockdevice_id: The unique identifier of the block device.
+    :ivar int size: The size of the block device.
+    :ivar bytes host: The IP address of the host to which the block device is
+        attached or ``None`` if it is currently unattached.
+    """
+    blockdevice_id = field(type=unicode, mandatory=True)
+    size = field(type=int, mandatory=True)
+    # XXX: Should be hostname, for consistency of is host a better name since
+    # we currently only expect IP addresses?
+    host = field(type=(bytes, type(None)), initial=None)
+
+
+def losetup_list_parse(output):
+    """
+    Parse the output of ``losetup --all`` which varies depending on the
+    privileges of the user.
+
+    :param unicode output: The output of ``losetup --all``.
+    :returns: A ``list`` of
+        2-tuple(FilePath(device_file), FilePath(backing_file))
+    """
+    devices = []
+    for line in output.splitlines():
+        parts = line.split(':', 2)
+        if len(parts) != 3:
+            continue
+        device_file, attributes, backing_file = parts
+        device_file = FilePath(device_file.strip())
+
+        # Trim everything from the first left bracket, skipping over the
+        # possible inode number which appears only when run as root.
+        left_bracket_offset = backing_file.find('(')
+        backing_file = backing_file[left_bracket_offset+1:]
+
+        # Trim everything from the right most right bracket
+        right_bracket_offset = backing_file.rfind(')')
+        backing_file = backing_file[:right_bracket_offset]
+
+        # Trim a possible embedded deleted flag
+        expected_suffix_list = ['(deleted)']
+        for suffix in expected_suffix_list:
+            offset = backing_file.rfind(suffix)
+            if offset > -1:
+                backing_file = backing_file[:offset]
+
+        # Remove the space that may have been between the path and the deleted
+        # flag.
+        backing_file = backing_file.rstrip()
+        backing_file = FilePath(backing_file)
+        devices.append((device_file, backing_file))
+    return devices
+
+
+def losetup_list():
+    """
+    List all the loopback devices on the system.
+
+    :returns: A ``list`` of
+        2-tuple(FilePath(device_file), FilePath(backing_file))
+    """
+    output = check_output(
+        ["losetup", "--all"]
+    ).decode('utf8')
+    return losetup_list_parse(output)
+
+
+def device_for_path(expected_backing_file):
+    """
+    :param FilePath backing_file: A path which may be associated with a
+        loopback device.
+    :returns: A ``FilePath`` to the loopback device if one is found, or
+        ``None`` if no device exists.
+    """
+    for device_file, backing_file in losetup_list():
+        if expected_backing_file == backing_file:
+            return device_file
+    return None
+
+
+@implementer(IBlockDeviceAPI)
+@attributes(['root_path'])
```

Comment:
It's not really a record class. I only used characteristic to avoid having to create an initialiser method.
But that's not a good reason, so I just added an __init__ method.